File: deprecated/examples/rdann.pl
#!/usr/bin/perl
#
# File: rdann.pl       I. Henry   March 29 2005
#
# Minimal WFDB annotator reader written in Perl, based on rdann.c
# 							   
# Copyright (C) 2005 Isaac C. Henry (ihenry@physionet.org)

package wfdb;
use wfdb;

if ( @ARGV < 4 ) {
    usage();
    exit;
} 

# parse the arguments
for ( $i=0; $i<@ARGV; $i++ ) {
    if ( $ARGV[$i] =~ m/\-r/ ) {
	$record = $ARGV[++$i];
    }
    elsif ( $ARGV[$i] =~ m/\-a/ ) {
	$annotator = $ARGV[++$i];
    }
    else {
	usage();
	exit;
    }
}

if ( ! $record || ! $annotator ) {
	usage();
	exit(2);
}

# set the sampling frequency
if ( ($sps = sampfreq($record)) < 0.) {
	setsampfreq($sps = $WFDB_DEFFREQ);
}

# get a new anninfo object
$aiarray = new wfdb::WFDB_AnninfoArray(1);

# set ai fields
$ai = $aiarray->getitem(0);
%$ai = ( name => $annotator,
	 stat => $WFDB_READ );
# this is also equivalent to
# $ai->{name} = $annotator;
# $ai->{stat} = $WFDB_READ;

# put the ai in the array
$aiarray->setitem(0,$ai);

# open the annotations
$result = annopen( $record, $aiarray->cast(), 1 );

if ( $result < 0 ) {
	usage();
	exit(2);
}

# get a new annotation object
$annot = new wfdb::WFDB_Annotation();

while ( getann(0, $annot) == 0 ) {    
    print mstimstr(-$annot->{time}), "\t",
    $annot->{time}, "\t",
    annstr( $annot->{anntyp} ), "\t",
    $annot->{subtyp}, "\t",
    $annot->{chan}, "\t",
    $annot->{num}, "\t",
    substr($annot->{aux}, 1), "\n";
}

wfdbquit();

sub usage {
    print STDERR "Usage: rdann.pl -r record -a annotator\n";
}

File: examples/rdann.py
#!/usr/bin/env python
#
# File: rdann.py       I. Henry      28 March 2005
#                      last revised: 13 January 2013
# 
# Minimal WFDB annotator reader written in Python, based on rdann.c
# 							   
# Copyright (C) 2013 Isaac C. Henry (ihenry42@gmail.org)
# This file is part of wfdb-swig.
#
# wfdb-swig is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# wfdb-swig is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with wfdb-swig.  If not, see <http://www.gnu.org/licenses/>.

import getopt, wfdb, sys

def main(argv):
    record = ''
    annotator = ''

    # parse the arguments
    try:
        opts, args = getopt.getopt(argv, "hr:a:", ["help"])
    except getopt.GetoptError:
        usage()
        sys.exit(2)

    for opt, arg in opts:
        if opt in ("-h", "--help"):
            usage()
            sys.exit()
        elif opt == '-r':
            record = arg
        elif opt == '-a':
            annotator = arg

    if len(record) == 0 or len(annotator) == 0:
        usage()
        sys.exit(2)
        
    # set the sampling frequency
    sps = wfdb.sampfreq(record)
    if sps < 0:
        sps = wfdb.WFDB_DEFFREQ
        wfdb.setsampfreq(sps)
            
    # get a new anninfo object
    ai = wfdb.WFDB_Anninfo()
    
    # set ai fields
    ai.name = annotator
    ai.stat = wfdb.WFDB_READ
    
    # open the annotation
    result = wfdb.annopen( record, ai, 1 )
    
    if result < 0:
        usage()
        sys.exit(2)
        
    # get a new annotation object
    annot = wfdb.WFDB_Annotation()
    
    while wfdb.getann(0, annot) == 0:

        # remove first char from aux string, if there is one
        if annot.aux is not None:
            aux = annot.aux[1:]
        else:
            aux = ""
                
        print "%s\t%d\t%s\t%d\t%d\t%d\t%s" % ( \
              wfdb.mstimstr(-annot.time), \
              annot.time, \
              wfdb.annstr(annot.anntyp),
              annot.subtyp,
              annot.chan, \
              annot.num, \
              aux)

    wfdb.wfdbquit()


def usage():
    print "Usage: rdann.py -r record -a annotator"

    
if __name__ == "__main__":
    main(sys.argv[1:])
